internal/generate/clusterserviceversion/clusterserviceversion.go

```diff
@@ -0,0 +1,240 @@
+func (g *Generator) generate() (*operatorsv1alpha1.ClusterServiceVersion, error) {
+	if g.getBase == nil {
+		return nil, genutil.InternalError("getBase must be set")
+	}
+
+	base, err := g.getBase()
+	if err != nil {
+		return nil, fmt.Errorf("error getting ClusterServiceVersion base: %v", err)
+	}
+
+	if err = g.updateVersions(base); err != nil {
+		return nil, err
+	}
+
+	if g.Collector != nil {
+		if err := applyTo(g.Collector, base); err != nil {
+			return nil, err
+		}
+	}
+
+	return base, nil
+}
+
+// makeCSVFileName returns a CSV file name containing name.
+func makeCSVFileName(name string) string {
+	return strings.ToLower(name) + csvYamlFileExt
+}
+
+// makeKustomizeBaseGetter returns a function that gets a kustomize-style base.
+func (g Generator) makeKustomizeBaseGetter(inputDir, apisDir string) getBaseFunc {
+	basePath := filepath.Join(inputDir, "bases", makeCSVFileName(g.OperatorName))
+	if genutil.IsNotExist(basePath) {
+		basePath = ""
+	}
+
+	return g.makeBaseGetter(basePath, apisDir)
+}
+
+// makeBaseGetter returns a function that gets a base from inputDir.
+// apisDir is used by getBaseFunc to populate base fields.
+func (g Generator) makeBaseGetter(basePath, apisDir string) getBaseFunc {
+	gvks := make([]schema.GroupVersionKind, len(g.config.Resources))
+	for i, gvk := range g.config.Resources {
+		gvks[i].Group = fmt.Sprintf("%s.%s", gvk.Group, g.config.Domain)
+		gvks[i].Version = gvk.Version
+		gvks[i].Kind = gvk.Kind
+	}
+
+	return func() (*operatorsv1alpha1.ClusterServiceVersion, error) {
+		b := bases.ClusterServiceVersion{
+			OperatorName: g.OperatorName,
+			OperatorType: g.OperatorType,
+			BasePath:     basePath,
+			APIsDir:      apisDir,
+			GVKs:         gvks,
+		}
+		return b.GetBase()
+	}
+}
+
+// updateVersions updates csv's version and data involving the version,
+// ex. ObjectMeta.Name, and place the old version in the `replaces` object,
+// if there is an old version to replace.
+func (g Generator) updateVersions(csv *operatorsv1alpha1.ClusterServiceVersion) (err error) {
+
+	oldVer, newVer := csv.Spec.Version.String(), g.Version
+	newName := genutil.MakeCSVName(g.OperatorName, newVer)
+	oldName := csv.GetName()
+
+	// A bundled CSV may not have a base containing the previous version to use,
+	// so use the current bundled CSV for version information.
+	if genutil.IsExist(g.bundledPath) {
+		existing, err := (bases.ClusterServiceVersion{BasePath: g.bundledPath}).GetBase()
+		if err != nil {
+			return fmt.Errorf("error reading existing ClusterServiceVersion: %v", err)
+		}
+		oldVer = existing.Spec.Version.String()
+		oldName = existing.GetName()
+	}
+
+	// If the new version is empty, either because a CSV is only being updated or
+	// a base was generated, no update is needed.
+	if newVer == "0.0.0" || newVer == "" {
+		return nil
+	}
+
+	// Set replaces by default.
+	// TODO: consider all possible CSV versioning schemes supported  by OLM.
+	if oldVer != "0.0.0" {
+		csv.Spec.Replaces = oldName
+	}
```

Comment:
if it has not in the api lib, could we not make a comment here and track this need for we use the lib check instead of impl the specific OLM validations here?